商品情報編集機能実装

shiki0501 opened the pull request:
# What
商品編集機能実装
# Why
出品した商品情報を編集できるようにすることで、間違えて入力してしまった情報を修正できるようにするため

コードレビューのほど、よろしくお願いいたします。

- ログイン状態の出品者は、商品情報編集ページに遷移できる動画
https://gyazo.com/785caf2f3f93ae7ddb370da299d5af39
- 正しく情報を記入すると、商品の情報を編集できる動画
https://gyazo.com/2597525ee4fc0c0ce41d7073c0e9d599
- 入力に問題のある状態では商品の情報は編集できず、エラーメッセージが出力される動画
https://gyazo.com/88c07644f14841f85aa6494415e608df
- 何も編集せずに更新をしても画像無しの商品にならない動画
https://gyazo.com/15e497453939b07ceb6004be531a4371
- ログイン状態のユーザーであっても、URLを直接入力して出品していない商品の商品情報編集ページへ遷移しようとすると、トップページに遷移する動画
https://gyazo.com/70d42b91e85fff9337d94e188cf6b9b7
- ログアウト状態のユーザーは、URLを直接入力して商品情報編集ページへ遷移しようとすると、ログインページに遷移する動画
https://gyazo.com/45d9b2b2b2f582d821f6ce104251bbcd
- ログイン状態の出品者であっても、URLを直接入力して自身の売却済み商品の商品情報編集ページへ遷移しようとすると、トップページに遷移する動画（現段階で商品購入機能の実装が済んでいる場合）
まだ未実装のため、省略
- 商品名やカテゴリーの情報など、すでに登録されている商品情報は商品情報編集画面を開いた時点で表示される動画（画像に関しては、表示されない状態で良い）
https://gyazo.com/69bdddf186208d992780b9d71a017bdf

shiki0501 committed:
商品編集画面遷移機能

editアクションの追加
edit.erb.htmlファイルの追加
routes設定
編集ボタンより編集ページへの遷移機能実装

shiki0501 committed:
編集機能実装

editアクションの@item定義
updateアクションの記述追加
edit.erb.htmlの各データ読み出し記述修正
routes updateアクション追加

shiki0501 committed:
未ログイン時の動作と戻るボタン

・未ログインに出品編集ページへ遷移するとログインを求められる
・編集ページの戻るボタンで詳細画面ページへ遷移する

shiki0501 committed:
エラーログ表示機能実装

エラーログ表示

shiki0501 committed:
rubocop使用

yoshi0809 commented on app/controllers/items_controller.rb:
```diff
@@ -1,5 +1,5 @@
 class ItemsController < ApplicationController
-  before_action :authenticate_user!, only: [:new, :create]
+  before_action :authenticate_user!, only: [:new, :create, :edit]
```
2行目
editアクションと同様にupdateアクションにも条件分岐を追加しましょう。

検証ツールなどを用いてリクエストを改ざんすることで、商品編集画面を経由せずにupdateアクションを動かすことが可能です。
そういった行為への対策として、updateアクションにもeditアクションと同様の条件分岐を加えることが必要となります。

yoshi0809 commented on app/controllers/items_controller.rb:
```diff
@@ -22,6 +18,24 @@ def create
   end
 
+  def show
+    @item = Item.find(params[:id])
+  end
+
+  def edit
+    @item = Item.find(params[:id])
+    redirect_to action: :index unless @item.user.id == current_user.id
+  end
+
+  def update
+    @item = Item.find(params[:id])
```
22、26、31行目
```rb
  @item = Item.find(params[:id])
```
こちらの記述は同じものですので、プライベートメソッド化してリファクタリングしてみましょう！

yoshi0809 commented on app/controllers/items_controller.rb:
```diff
@@ -22,6 +18,24 @@ def create
+
+  def edit
+    @item = Item.find(params[:id])
+    redirect_to action: :index unless @item.user.id == current_user.id
```
27行目
こちらの条件式もプライベートメソッド化して、updateアクションにも適用するようにしておきましょう。

【補足】
上記コメントの
```rb
@item = Item.find(params[:id])
```
とは別のプライベートメソッドとして定義した方が汎用性が高まります。

yoshi0809 commented on app/controllers/items_controller.rb:
```diff
@@ -22,6 +18,24 @@ def create
+  def update
+    @item = Item.find(params[:id])
+    if @item.update(item_params)
+      redirect_to item_path
```
33行目

```rb
item_path(@item)
```
としておきましょう。

【理由】
item_pathのURIを見ると`:id`があります。
引数として`:id`を渡さない場合railsが自動補間してくれますが、ネストをしている場合や記述ミスにより意図しない`:id`を補間してしまう場合があります。
そのため明示的に引数を渡しておくことで、意図しない挙動を回避することができます。

yoshi0809 commented on app/views/items/edit.html.erb:
```diff
@@ -0,0 +1,159 @@
+    <%# 下部ボタン %>
+    <div class="sell-btn-contents">
+      <%= f.submit "変更する" ,class:"sell-btn" %>
+      <%=link_to 'もどる', item_path, class:"back-btn" %>
```
142行目
こちらも
```rb
item_path(@item)
```
としておきましょう。

yoshi0809 commented on app/views/items/show.html.erb:
```diff
@@ -25,7 +25,7 @@
     <% if user_signed_in? %>
        <% if current_user.id == @item.user_id %>
-      <%= link_to "商品の編集", "#", method: :get, class: "item-red-btn" %>
+      <%= link_to "商品の編集", edit_item_path , method: :get, class: "item-red-btn" %>
```
28行目
こちらも
```rb
edit_item_path(@item)
```
としておきましょう。

yoshi0809 reviewed (commented):
レビュー依頼ありがとうございます。
修正点確認させていただきました。
追加でコメントいたしましたので、ご確認お願いいたします。

ご確認＆ご修正いただけましたら、コミット→プッシュ→フォームで再提出をお願いします！
ご不明点や疑問点ございましたら、コネクトメンターにお気軽にご相談ください！
どうぞよろしくお願いいたします。

shiki0501 committed:
コードレビュー後の修正

controller
・ログインしていない時はupdateページに遷移できない
・記述のプライベートメソッド化によるリファクタリング
erb
・route設定の詳細化
明示的に引数を渡しておくことで、意図しない挙動を回避することができる

shiki0501 committed:
コードレビュー後の修正

セキュリティーの強化
ログインしたユーザーが投稿したユーザーを等しくないと編集ページに遷移できないかつ編集記述を上書きできない（検証ツールなどにて不正操作して編集されることを防ぐ）

div-reviewer reviewed (commented):
shiki0501様、学習お疲れ様です！
お忙しい中修正と再レビュー依頼ありがとうございました。

前回コメントがあった6箇所が綺麗に修正できていますね！
素晴らしいです！

こちら`LGTM`とさせていただきます！実装お疲れさまでした。
次のブランチを切る前にMergeとPullを忘れないようお願いします！

次は「商品削除機能」かと思われます！引き続き頑張りましょう！応援しております！
